### PR TITLE
#166880073 User should be able to signup.

### PR DESCRIPTION
**Pull request by frankopkusianwar**
#### What does this PR do?

- Adds signup functionality to the application

#### Description of Task to be completed?

- Add a signup form with username input, email input, and password input fields.
- Pick data from the signup form and make a request to the `ah-backend-API` to register a user.

#### How should this be manually tested?

- visit ` https://ah-frontend-space-stagin-pr-17.herokuapp.com/signup `to view the signup page and test the app

#### What are the relevant pivotal tracker stories?

[#166880073](https://www.pivotaltracker.com/story/show/166880073)

#### Screenshots (if appropriate)

scrshot

### CODE REVIEW

File: src/assets/css/signup.scss [hunk 1]
```diff
@@ -0,0 +1,45 @@
+.signuppage{
+    border: 0px solid black;
+    padding: 5%;
+}
+
+.form-contents{
+    border: 1px solid #00bfff;
```
Comment (NicholusMuwonge) [inline src/assets/css/signup.scss]: Please kindly format this file with indentation of 2 spaces and same for every other .scss file

**Review by NicholusMuwonge — CHANGES_REQUESTED** [review]
Kindly react to this feedback as I pull this branch to go through it a little bit more

File: src/__test__/actions/signupActions.test.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import moxios from 'moxios';
+
+import signupAction from '../../actions/signupAction';
+import { REGISTER_USER, REGISTER_USER_SUCESS, REGISTER_USER_FAILURE } from '../../actions/types';
```
Comment (NicholusMuwonge) [inline src/__test__/actions/signupActions.test.js]: If REGISTER_USER_FAILURE isn't being used please remove it

File: src/assets/css/signup.scss (hunk 1 above)
Comment (frankopkusianwar) [reply src/assets/css/signup.scss]: let me work on this @NicholusMuwonge

File: src/__test__/reducers/signupreducer.test.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { isTSAnyKeyword, exportAllDeclaration } from '@babel/types';
```
Comment (NicholusMuwonge) [inline src/__test__/reducers/signupreducer.test.js]: kindly remove this if it's not being used

File: src/__test__/reducers/signupreducer.test.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import signupReducer from '../../reducers/signupReducer';
+import { registerUser, registerUserSucess, registerUserfailure } from '../../actions/signupAction';
```
Comment (NicholusMuwonge) [inline src/__test__/reducers/signupreducer.test.js]: This too, it seems to be not used at all. kindly take it out.

File: src/actions/signupAction.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable jsx-a11y/anchor-is-valid */
```
Comment (NicholusMuwonge) [inline src/actions/signupAction.js]: Kindly take out this line please

**Comment by NicholusMuwonge** [comment]
- I kindly recommend that you lose the picture on the left of the login, center the signup title. 
- Kindly also remove the focus that appears when someone clicks on the button.
- Add a loader to let the user know that something is changed.
- If you could remove the inner input field shadows.

File: src/assets/css/signup.scss (hunk 1 above)
Comment (frankopkusianwar) [reply src/assets/css/signup.scss]: @NicholusMuwonge this has been resolved

File: src/__test__/actions/signupActions.test.js (hunk 2 above)
Comment (frankopkusianwar) [reply src/__test__/actions/signupActions.test.js]: @NicholusMuwonge this has been worked on

File: src/__test__/reducers/signupreducer.test.js (hunk 3 above)
Comment (frankopkusianwar) [reply src/__test__/reducers/signupreducer.test.js]: @NicholusMuwonge this has been worked on

File: src/__test__/reducers/signupreducer.test.js (hunk 4 above)
Comment (frankopkusianwar) [reply src/__test__/reducers/signupreducer.test.js]: @NicholusMuwonge this has been worked on

**Review by NicholusMuwonge — APPROVED** [review]
works as expected and thanks for reacting to the feedback I left on your PR

**Review by sebbss — APPROVED** [review]
looks good to me

**Review by richien — APPROVED** [review]
This works as described. Good work @frankopkusianwar

**Review by MuhanguziDavid — CHANGES_REQUESTED** [review]
Nice work on signup @frankopkusianwar . However, please punctuate the success message that is displayed on successful signup or find a better way to state it. Try something like `Verification email has been sent to <email address>, please check your email to verify.`

Screenshot 2019-07-18 at 20 13 41

**Review by MuhanguziDavid — COMMENTED** [review]
I also encountered this! Dismiss all previous toasts i.e. `toast.dismiss()` as the first statement in your action (or wherever you rendered the toasts from) to avoid new toasts from being rendered while old ones are still existent.

Screenshot 2019-07-18 at 20 14 25

**Review by MuhanguziDavid — COMMENTED** [review]
I see the URL pathname as `/login` i.e. login page but get the message `Ooops! Page not found`.
The login page exists, please ensure that a user is redirected to the login page on successful signup. 

Screenshot 2019-07-18 at 20 19 19

**Comment by MuhanguziDavid** [comment]
Please update the image in your PR description

File: .eslintrc.json [hunk 6]
```diff
@@ -1,43 +1,47 @@
...
+      "react/prefer-stateless-function": 0,
+      "react/jsx-filename-extension": [
+          1,
+          {
+              "extensions": [".js", ".jsx"]
+          }
+      ]
   }
-}
+}
```
Comment (MuhanguziDavid) [inline .eslintrc.json]: please add a line at the end of the file

File: .travis.yml [hunk 7]
```diff
@@ -13,4 +13,4 @@ script:
   - npm test -- -u
 after_script:
   - npm run coveralls
-
+
```
Comment (MuhanguziDavid) [inline .travis.yml]: new line at the end of the file

File: src/__test__/components/signup/SignupForm.test.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    it('should map redux state values to component props', () => {
+      const state = {
+        signupReducer: { isSigningUp: false },
+      };
+      expect(mapStateToProps(state)).toEqual({ isSigningUp: false });
+    });
+    it('should redirect on successful signup', () => {
+      wrapper.setProps({ ...nextProps });
+      expect(props.history.push).toBeCalledWith('/login');
+    });
```
Comment (MuhanguziDavid) [inline src/__test__/components/signup/SignupForm.test.js:76]: Nice work with the redirection test. Please test for not redirecting when signup is unsuccessful i.e. `isValid: false`

**Review by MuhanguziDavid — APPROVED** [review]
LGTM

**Pull request merged**